authservice/auth_service.go

```diff
@@ -329,47 +571,231 @@ func (r *RemoteAuthDataLoaderImpl) GetAccessToken() error {
 		return fmt.Errorf("error getting access token: %d - %s", resp.StatusCode, string(body))
 	}
 
-	err = json.Unmarshal(body, &r.accessToken)
+	var accessToken AccessToken
+	err = json.Unmarshal(body, &accessToken)
 	if err != nil {
 		return fmt.Errorf("error on unmarshal access token response: %v", err)
 	}
 
+	r.accessTokens.Store(AppOrgPair{AppID: appID, OrgID: orgID}, accessToken)
+
 	return nil
 }
 
-// GetDeletedAccounts implements AuthDataLoader interface
-func (r *RemoteAuthDataLoaderImpl) GetDeletedAccounts() ([]string, error) {
-	accountIDs, err := r.requestDeletedAccounts()
+// GetAccessTokens implements AuthDataLoader interface
+func (r *RemoteAuthDataLoaderImpl) GetAccessTokens() error {
+	if r.config == nil {
+		return fmt.Errorf("auth data loader is not configured")
+	}
+	req, err := r.config.BuildAccessTokensRequest()
+	if err != nil {
+		return fmt.Errorf("error creating access tokens request: %v", err)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error requesting access tokens: %v", err)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("error reading body of access tokens response: %v", err)
+	}
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("error getting access tokens: %d - %s", resp.StatusCode, string(body))
+	}
+
+	var accessTokens []accessTokensResponse
+	err = json.Unmarshal(body, &accessTokens)
+	if err != nil {
+		return fmt.Errorf("error on unmarshal access tokens response: %v", err)
+	}
+
+	r.accessTokens = &sync.Map{}
+	r.pairsLock.Lock()
+	defer r.pairsLock.Unlock()
+
+	r.appOrgPairs = make([]AppOrgPair, len(accessTokens))
+	for i, res := range accessTokens {
+		pair := appOrgPairFromResponse(appOrgPairResponse{AppID: res.AppID, OrgID: res.OrgID})
+
+		r.appOrgPairs[i] = pair
+		r.accessTokens.Store(pair, res.AccessToken)
+	}
+
+	return nil
+}
+
+// MakeRequest implements AuthDataLoader interface
+func (r *RemoteAuthDataLoaderImpl) MakeRequest(requestFunc func(AppOrgPair, AccessToken) (interface{}, error), appOrgPair AppOrgPair, retryString string, updateTokenIfNeeded bool) (interface{}, error) {
+	token, appOrgPair, ok := r.isAccessGranted(appOrgPair)
+	if !ok {
+		return nil, errors.New("access not granted")
+	}
+
+	data, err := requestFunc(appOrgPair, token)
 	if err != nil {
-		if strings.HasPrefix(err.Error(), "error getting deleted accounts: 401") {
+		if updateTokenIfNeeded && strings.HasPrefix(err.Error(), retryString) {
 			// access token may have expired, so get a new one and try once more
-			tokenErr := r.GetAccessToken()
-			if tokenErr != nil {
-				return nil, fmt.Errorf("error getting new access token - %v - after %v", tokenErr, err)
+			token, updateErr := r.updateAccessToken(appOrgPair)
```

Comment:
I think we should probably just refresh all tokens here since it is likely that the others would have expired as well.